File: backend/app/repositories/feedback.py
from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)


class DuplicateFeedbackSurveyError(Exception):
    pass


class FeedbackRepository:
    def __init__(self, *, supabase: Any):
        self._supa = supabase

    @staticmethod
    def _normalize_ip(ip: str) -> str:
        return (ip or "unknown")[:100]

    @staticmethod
    def _is_duplicate_key_error(exc: Exception) -> bool:
        text = str(exc or "").lower()
        return "23505" in text or "duplicate key" in text or ("unique" in text and "feedback_surveys" in text)

    def insert_feedback(
        self,
        *,
        ip: str,
        request_id: str | None,
        message: str,
        email: str | None,
        path: str,
        user_agent: str | None,
        referrer: str | None,
    ) -> None:
        payload = {
            "ip": self._normalize_ip(ip),
            "request_id": (request_id or "").strip()[:100] or None,
            "message": (message or "").strip()[:10_000],
            "email": (email or "").strip()[:320] or None,
            "path": (path or "").strip()[:500],
            "user_agent": (user_agent or "").strip()[:500] or None,
            "referrer": (referrer or "").strip()[:500] or None,
        }

        try:
            self._supa.table("feedback").insert(payload).execute()
        except Exception:
            # Best-effort; do not break UI if DB is missing.
            logger.exception("Failed to insert feedback")

    def has_feedback_survey_for_ip(self, *, ip: str) -> bool:
        response = self._supa.table("feedback_surveys").select("id").eq("ip", self._normalize_ip(ip)).limit(1).execute()
        rows = response.data or []
        return any(isinstance(row, dict) and row.get("id") is not None for row in rows)

    def insert_feedback_survey(
        self,
        *,
        ip: str,
        request_id: str | None,
        subscription_intent: str,
        fair_price_monthly: float | None,
        usage_frequency: str,
        primary_market_focus: str,
        discovery_source: str,
        trust_score: int,
        referral_likelihood: int,
        most_wanted_feature: str,
        must_improve_before_pay: str,
        ideal_alert_channel: str | None,
        additional_notes: str | None,
        web_helpful: str | None,
        email: str | None,
        path: str,
        user_agent: str | None,
        referrer: str | None,
    ) -> None:
        payload = {
            "ip": self._normalize_ip(ip),
            "request_id": (request_id or "").strip()[:100] or None,
            "subscription_intent": (subscription_intent or "").strip()[:20],
            "fair_price_monthly": round(float(fair_price_monthly), 2) if fair_price_monthly is not None else None,
            "usage_frequency": (usage_frequency or "").strip()[:20],
            "primary_market_focus": (primary_market_focus or "").strip()[:40],
            "discovery_source": (discovery_source or "").strip()[:40],
            "trust_score": int(trust_score),
            "referral_likelihood": int(referral_likelihood),
            "most_wanted_feature": (most_wanted_feature or "").strip()[:500],
            "must_improve_before_pay": (must_improve_before_pay or "").strip()[:1_000],
            "ideal_alert_channel": (ideal_alert_channel or "").strip()[:200] or None,
            "additional_notes": (additional_notes or "").strip()[:4_000] or None,
            "web_helpful": (web_helpful or "").strip()[:20] or None,
            "email": (email or "").strip()[:320] or None,
            "path": (path or "").strip()[:500],
            "user_agent": (user_agent or "").strip()[:500] or None,
            "referrer": (referrer or "").strip()[:500] or None,
        }

        try:
            self._supa.table("feedback_surveys").insert(payload).execute()
        except Exception as exc:
            if self._is_duplicate_key_error(exc):
                raise DuplicateFeedbackSurveyError() from exc
            logger.exception("Failed to insert feedback survey")
            raise

File: backend/internal/repo/feedback.go
package repo

import (
	"fmt"
	"log/slog"
	"strings"

	"yunews/backend/internal/supabase"
)

// ErrDuplicateFeedbackSurvey is returned when a duplicate unique constraint is violated.
var ErrDuplicateFeedbackSurvey = fmt.Errorf("duplicate feedback survey")

// FeedbackRepository handles feedback and feedback_surveys tables.
type FeedbackRepository struct {
	db *supabase.Client
}

// NewFeedbackRepository creates a new FeedbackRepository.
func NewFeedbackRepository(db *supabase.Client) *FeedbackRepository {
	return &FeedbackRepository{db: db}
}

// InsertFeedback inserts a feedback row. Errors are logged but not returned (best-effort).
func (r *FeedbackRepository) InsertFeedback(ip, requestID, message string, email *string, path, userAgent, referrer *string) {
	payload := map[string]interface{}{
		"ip":          truncate(normalizeIP(ip), 100),
		"request_id":  truncateOrNil(requestID, 100),
		"message":     truncate(strings.TrimSpace(message), 10000),
		"email":       truncateOrNilPtr(email, 320),
		"path":        truncate(strDefault(path, "/"), 500),
		"user_agent":  truncateOrNilPtr(userAgent, 500),
		"referrer":    truncateOrNilPtr(referrer, 500),
	}
	if err := r.db.Insert("feedback", payload); err != nil {
		slog.Warn("failed to insert feedback", "error", err)
	}
}

// HasFeedbackSurveyForIP returns true if the IP has already submitted a survey.
func (r *FeedbackRepository) HasFeedbackSurveyForIP(ip string) bool {
	var rows []map[string]interface{}
	err := r.db.From("feedback_surveys").
		Select("id").
		Eq("ip", normalizeIP(ip)).
		Limit(1).
		Execute(&rows)
	if err != nil {
		slog.Warn("failed to check feedback survey for ip", "error", err)
		return false
	}
	for _, row := range rows {
		if row["id"] != nil {
			return true
		}
	}
	return false
}

// InsertFeedbackSurvey inserts a feedback survey row.
// Returns ErrDuplicateFeedbackSurvey on unique-constraint violation.
func (r *FeedbackRepository) InsertFeedbackSurvey(params FeedbackSurveyParams) error {
	payload := map[string]interface{}{
		"ip":                    truncate(normalizeIP(params.IP), 100),
		"request_id":            truncateOrNil(params.RequestID, 100),
		"subscription_intent":   truncate(params.SubscriptionIntent, 20),
		"fair_price_monthly":    params.FairPriceMonthly,
		"usage_frequency":       truncate(params.UsageFrequency, 20),
		"primary_market_focus":  truncate(params.PrimaryMarketFocus, 40),
		"discovery_source":      truncate(params.DiscoverySource, 40),
		"trust_score":           params.TrustScore,
		"referral_likelihood":   params.ReferralLikelihood,
		"most_wanted_feature":   truncate(params.MostWantedFeature, 500),
		"must_improve_before_pay": truncate(params.MustImproveBeforePay, 1000),
		"ideal_alert_channel":   truncateOrNil(params.IdealAlertChannel, 200),
		"additional_notes":      truncateOrNil(params.AdditionalNotes, 4000),
		"web_helpful":           truncateOrNil(params.WebHelpful, 20),
		"email":                 truncateOrNilPtr(params.Email, 320),
		"path":                  truncate(strDefault(&params.Path, "/"), 500),
		"user_agent":            truncateOrNilPtr(params.UserAgent, 500),
		"referrer":              truncateOrNilPtr(params.Referrer, 500),
	}

	if err := r.db.Insert("feedback_surveys", payload); err != nil {
		if isDuplicateKeyError(err) {
			return ErrDuplicateFeedbackSurvey
		}
		return err
	}
	return nil
}

// FeedbackSurveyParams holds all fields for a feedback survey insert.
type FeedbackSurveyParams struct {
	IP                   string
	RequestID            string
	SubscriptionIntent   string
	FairPriceMonthly     *float64
	UsageFrequency       string
	PrimaryMarketFocus   string
	DiscoverySource      string
	TrustScore           int
	ReferralLikelihood   int
	MostWantedFeature    string
	MustImproveBeforePay string
	IdealAlertChannel    string
	AdditionalNotes      string
	WebHelpful           string
	Email                *string
	Path                 string
	UserAgent            *string
	Referrer             *string
}

func normalizeIP(ip string) string {
	ip = strings.TrimSpace(ip)
	if ip == "" {
		return "unknown"
	}
	if len(ip) > 100 {
		return ip[:100]
	}
	return ip
}

func truncate(s string, maxLen int) string {
	s = strings.TrimSpace(s)
	if len(s) > maxLen {
		return s[:maxLen]
	}
	return s
}

func truncateOrNil(s string, maxLen int) interface{} {
	s = strings.TrimSpace(s)
	if s == "" {
		return nil
	}
	if len(s) > maxLen {
		s = s[:maxLen]
	}
	return s
}

func truncateOrNilPtr(s *string, maxLen int) interface{} {
	if s == nil {
		return nil
	}
	return truncateOrNil(*s, maxLen)
}

func strDefault(s *string, def string) *string {
	if s == nil || strings.TrimSpace(*s) == "" {
		return &def
	}
	return s
}

func isDuplicateKeyError(err error) bool {
	if err == nil {
		return false
	}
	msg := strings.ToLower(err.Error())
	return strings.Contains(msg, "23505") || strings.Contains(msg, "duplicate key") ||
		(strings.Contains(msg, "unique") && strings.Contains(msg, "feedback_surveys"))
}
